### ISSUE TITLE
Grpc.Tools version 2.55.1 not resolved to specific version error

### ISSUE THREAD

**Issue by Kubana** [CycloneDX/cyclonedx-dotnet#894]
Couple of days ago I started getting this exception when generating SBOM from one of our projects.

Dependency (Grpc.Tools) with version range ([2.55.1, )) referenced by (Name:Grpc.AspNetCore Version:2.55.0) did not resolve to a specific version.
Unable to locate valid bom ref for Grpc.Tools [2.55.1, )

What is the issue here?

**Comment by mtsfoni** [comment]
Are you using the current version of cyclonedx-dotnet? I hoped we got rid of that bug.

Do you maybe have a different version of grpc.tools referenced directly?

**Comment by Kubana** [comment]
I think I have but let me doublecheck

**Comment by Kubana** [comment]
Currently running the dotnet tool CycloneDX version 3.0.8

**Comment by mtsfoni** [comment]
With what argument do you call the tool?
Normaly a single project.assets.json should be unambiguous.
This problem can arise, if you have a solution where different projects refer to different versions of a library as the tool doesn't know which version to use.

I want to change the tool to just add all versions of that library in case, but I have no free capacity.

A workaround can be, to generate the sboms by refering the "root-component"-projects (meaning your .exe / service-backend ). I recommend doing it that way for increased accuracy anyway and handle all my projects like that. (I generate a frontend and backend bom seperately then)

**Comment by Kubana** [comment]
dotnet-CycloneDX.exe {projectName}.csproj -o CycloneDXOutputs -fn {projectName}.xml -dpr

**Comment by Kubana** [comment]
Problem persists even in version 4.0.0

**Comment by mtsfoni** [comment]
Hey,

is this still a problem for you? I just tried but I just cannot get it reproduced

Could you describe the dependencies between the projects and exactly which version of Grpc.AspNetCore/Tools is referenced by which project?

**Comment by Falco20019** [comment]
I have the same issue. I am currently looking to create a regression test for this and maybe also a fix.

**Comment by Falco20019** [comment]
@mtsfoni This is exactly what the test `ProjectAssetsFileServiceTests.GetDotnetDependencys_MissingResolvedPackageVersion` tries. While the assets part is fine with keeping the dependency, `NugetV3Service.GetComponentAsync` will fail to resolve it and fail the generation. I resolved it with the linked PR to some extend that will work for my use-case and also this one. It will still fail more exotic ones like those (basically any that excludes the min value):
* (1.0,)
* (,1.0]
* (,1.0)
* (1.0,2.0)

I also documented how this could be resolved with a bit more code to work properly, but usually it's easy to avoid and adjust the `csproj` to include existing min versions. Especially since the local nuget cache is not resolved through the NuGet API but manually, checking the locally cached versions isn't that easy and would need bigger reworks. It's possible as I already have an internal tool that does this properly and supports multiple package sources in parallel.

**Comment by mtsfoni** [comment]
@Falco20019: The issue I’m running into is that the test `ProjectAssetsFileServiceTests.GetDotnetDependencys_MissingResolvedPackageVersion` mocks a `project.assets.json` file where dependencies under `targets/library/dependencies` have a version range. But I haven’t been able to reproduce this behaviour in a real project.

As far as I understand, the `project.assets.json` file’s purpose is to determine specific versions during NuGet restore, so I wouldn’t expect it to contain version ranges. But maybe I’m missing some constellation?

Would you be able to produce a project-file that results in this behaviour? Right now, it seems to me the test is simulating a case that doesn’t naturally happen when building a solution, but I want to fully understand the issue.

**Comment by Falco20019** [comment]
@mtsfoni You just need to reference a nupkg that has a version range as published dependency. One public case is `Google.Api.CommonProtos@2.16.0` which has a dependency onto `Google.Protobuf` as `[3.28.2, 4.0.0)`. That's a pretty common feature and also nuget.org shows it properly: https://www.nuget.org/packages/Google.Api.CommonProtos/2.16.0#dependencies-body-tab

We have LOTS of those in our projects but many are internal libraries. You usually only set the max. if you know it's not running with a higher version OR if you thightly review it (like we do internally) and want to whitelist them to avoid end-users accidentally shooting themselves in the foot...

The `project.assets.json` will look like:
```json
{
  "version": 3,
  "targets": {
    "net8.0": {
      "Google.Api.CommonProtos/2.16.0": {
        "type": "package",
        "dependencies": {
          "Google.Protobuf": "[3.28.2, 4.0.0)"
        }
      }
},
  "projectFileDependencyGroups": {
    "net8.0": [
      "Google.Protobuf >= 3.28.2 < 4.0.0"
    ]
  }
  "project": {
    "frameworks": {
      "net8.0": {
        "dependencies": {
          "Google.Protobuf": {
            "target": "Package",
            "version": "[3.28.2, 4.0.0)",
            "versionCentrallyManaged": true // Only if CPM was used, otherwise missing
          }
        },
        "centralPackageVersions": { // Only if CPM was used, otherwise probably empty
          "Google.Protobuf": "[3.28.2, 4.0.0)"
        }
      }
    }
  }
}
```

If you would reference `Google.Protobuf` i.e. `Google.Protobuf@3.28.3` in the `csproj`, the `targets/Google.Api.CommonProtos/2.16.0/dependencies/Google.Protobuf` will still say `[3.28.2, 4.0.0)` while the `projectFileDependencyGroups/net8.0/Google.Protobuf` will change to `>= 3.28.3`. `project/frameworks/net8.0/dependencies/Google.Protobuf/version` is changed to `[3.28.3, )` and `project/frameworks/net8.0/centralPackageVersions/Google.Protobuf` is `3.28.3`.

**Comment by mtsfoni** [comment]
So I managed to get the error with this constellation. I assume this is basically your problem?

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFramework>net9.0</TargetFramework>
	</PropertyGroup>
	<ItemGroup>
	  <ProjectReference Include="..\ProjectB\ProjectB.csproj" />
	</ItemGroup>
</Project>
---
<!--ProjectB-->
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net9.0</TargetFramework>
  </PropertyGroup>

	<ItemGroup>		
		<PackageReference Include="Google.Api.CommonProtos" Version="2.16.0" />
		<PackageReference Include="Google.Protobuf" Version="3.28.2">
			<PrivateAssets>all</PrivateAssets>
		</PackageReference>
	</ItemGroup>
</Project>
---
> dotnet-cyclonedx .\ProjectA\ProjectA.csproj                                          1.697s

Found the following local nuget package cache locations:
    C:\Users\user\.nuget\packages\
Scanning at C:\Users\user\AppData\Local\Temp\GrpcRepro\ProjectA\ProjectA.csproj

» Analyzing: C:\Users\user\AppData\Local\Temp\GrpcRepro\ProjectA\ProjectA.csproj
  Attempting to restore packages
  Packages restored
**Dependency (Google.Protobuf) with version range ([3.28.2, 4.0.0)) referenced by (Name:Google.Api.CommonProtos Version:2.16.0) did not resolve to a specific version.**
Found 2 packages
Unable to locate valid bom ref for Google.Protobuf [3.28.2, 4.0.0)
```

**Comment by Falco20019** [comment]
Yes, and even when using the -rs flag it's not able to succeed. #929 solves it at least with the flag and also fixes some other problems of that mode.

**Comment by Falco20019** [comment]
I think the PR is about to get closed anyway if you want to avoid the recursive flag in those cases. So it will be more about seeing what other things I stumbled over while trying to solve it that raised some questions.

I can also open separate issues for each of those if you like to have a better picture and discuss what's the expected behavior first.

**Comment by mtsfoni** [comment]
> I can also open separate issues for each of those if you like to have a better picture and discuss what's the expected behavior first.

I will go over it and comment on parts that we can use. Recursive probably has some potential for improvements as well.

**Comment by Kubana** [comment]
Thank you issue is gone, sbom generation works.

![Image](https://github.com/user-attachments/assets/a7e1cbe2-ff56-496c-897c-3260c6ef0c93)
